### DISCUSSION TITLE
[Python] Closing star/cif files after opening

### DISCUSSION THREAD

**Discussion by paulkorir** [project-gemmi/gemmi#282] [Q&A] [answered]
This is in reference to the examples in the documentation e.g. https://gemmi.readthedocs.io/en/latest/cif.html#python. Is there a way to close files after they are opened that is not documented? I've browsed the source code and see that the Python wrappers directly reference the C++, which has no explicit file closing method. The danger is that it is likely to end up with too many opened handles. 

Kindly assist.

Paul K

**Comment by paulkorir** [comment]
I propose that the best way to handle this is to use the `cif.read_string(...)` method to a file handle [within a context manager](https://docs.python.org/3/reference/datamodel.html#with-statement-context-managers).

```python
from gemmi import cif

with open("file.star") as f:
    doc = cif.read_string(f.read())
```
which guarantees that the file will be closed.

**Comment by wojdyr** [comment, marked as answer]
The functions in Python bindings that are reading files,  they open a file, read the content, and close the file. They should be more efficient than reading the file as string in Python.

**Reply by paulkorir** [reply]
Great! It wasn't obvious to me. It would be a good titbit to include in the documentation lest anyone ask this again.
